DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+}
+```
+
+The symbol `X` is looked up per the current language rules in the working code above.
+
+If two or more imports within the same `with` declaration define the same name, name resolution across these is the same as if the imports were top-level.
+
+If a module defines a symbol at top level and then imports a module, lookup proceeds similarly with local imports. Consider:
+
+```d
+int writeln;
+with (import std.stdio)
+void main() { writeln("hello, world"); }
+```
+This code is in error because `writeln` has type int. However, the following code is correct because it specifies the symbol name explicitly:
+
+```d
+int writeln;
+with (import std.stdio : writeln)
+void main() { writeln("hello, world"); }
+```
+
+### Examples
+
+Below are a few examples taken from the standard library:
+
+```d
+with (import std.meta, std.range, std.traits)
+auto uninitializedArray(T, I...)(I sizes) nothrow @system
+if (isDynamicArray!T && allSatisfy!(isIntegral, I) &&
+    hasIndirections!(ElementEncodingType!T))
+{
+    ...
+}
+```
+
+Alternatively, the declaration may specify the exact symbols needed by using multiple imports:
+
+```d
+with (import std.meta : allSatisfy)
+with (import std.range : ElementEncodingType)
+with (import std.traits : hasIndirections, isDynamicArray, isIntegral)
+auto uninitializedArray(T, I...)(I sizes) nothrow @system
+if (isDynamicArray!T && allSatisfy!(isIntegral, I) &&
+    hasIndirections!(ElementEncodingType!T))
+{
+    ...
+}
+```
+
+#### Alternative: Lazy Imports
+
+Assume all `import`s are lazy without any change in the language. (This has already been implemented in the SDC compiler.) The way the scheme works, all `import`s seen are not yet executed but instead saved in a list of package/module names. Following that, the actual imports are triggered by one of two situations.
+
+First, consider the current module looks up a fully specified name:
+
+```d
+import module_a, module_b;
+void fun(T)(T value) if (module_a.condition!T)
+{
+    return module_b.process(value);
+}
+void fun(T)(T value) if (is(T == int)) { ... }
+```
+
+In this situation:
+
+* If `fun` is never looked up, neither `module_a` nor `module_b` needs to be loaded.
+* If `fun(42)` is used, even though the second overload is a match, then `module_a` must be loaded in order to ensure that `module_a.condition!int` is false so as to avoid ambiguity.
+* If `fun` is called with a non-`int` value, `module_a` is loaded to evaluate the template constraint. If the constraint is true, then `module_b` is also loaded so as to look up `process`.
+
+Let us note that full specification of symbols used may be enabled with ease by using the `static import` feature. We will henceforth refer to this setup as "the `static import` setup".
+
+Second, consider the situation (arguably more frequent in today's D code) when the current module does not fully specify names used. Instead, it imports the appropriate modules and relies on lookup to resolve symbols appropriately:
```

Comment:
That argumentation doesn't make sense, `static imports` would solve the problem, but people use unqualified names everywhere, so let's instead use `with (import module_a)`.